src/logging/LazyLoggerFactory.ts

```diff
@@ -1,45 +1,110 @@
-import { LazyLogger } from './LazyLogger';
-import type { Logger } from './Logger';
+import { Logger } from './Logger';
 import type { LoggerFactory } from './LoggerFactory';
+import type { LogLevel } from './LogLevel';
 
 /**
- * Wraps over another {@link LoggerFactory} that can be set lazily.
- * This is a singleton class, for which the instance can be retrieved using {@link LazyLoggerFactory.getInstance}.
- *
- * Loggers can safely be created before a {@link LoggerFactory} is set.
- * But an error will be thrown if {@link Logger.log} is invoked before a {@link LoggerFactory} is set.
- *
- * This creates instances of {@link LazyLogger}.
+ * Wraps around another {@link Logger} that can be set lazily.
  */
-export class LazyLoggerFactory implements LoggerFactory {
-  private static readonly instance = new LazyLoggerFactory();
-
-  private ploggerFactory: LoggerFactory | undefined;
+class WrappingLogger extends Logger {
+  public logger!: Logger;
```

Comment:
```suggestion
  public logger: Logger;
  
  public constructor(logger: Logger) {
    this.logger = logger;
  }
```

Or even more specifically, a `SimpleLogger` which only has the `log` function and is extended by `Logger` so no casting is required further down.

Reply:
Your suggested code won't work. Have a closer look at https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1248/files/a9394c9f874f86789ef168d64137e40542ff34c2#diff-b5cc5142b313b33feb8061e400d84b72b7528fe75df347350e50c32ea4342e72R30-R33: the argument you want to pass to the constructor depends on the return value of that constructor.

An alternate version of the code read:

```typescript
    const factory = this;
    const wrapper = new WrappingLogger();
    wrapper.logger = new BufferingLogger extends Logger {
      log(level: LogLevel, message: string, meta: any) {
        factory.bufferLogEntry(wrapper, level, message, meta)
      }
    }();
```
but the final version I landed on was easier.

I could do `SimpleLogger` but I don't see the advantage for an internal class. It introduces a new concept just for the sake of avoiding the cast, leaving one wondering what the new concept is and why it is used.

Reply:
> the argument you want to pass to the constructor here actually depends on the return value of that constructor.

True, I missed that. I just really want to prevent usage of operators that sidestep the typings as much as possible.

Reply:
> but I don't see the advantage for an internal class.

The main goal of my comment was type safety. The suggestion of `SimpleLogger` was to have a simply interface (similar to `Constructor` in `LogUtil` for example) so there was no need to cast to `any` when creating the logger below. It's very minor in this case but in my opinion the small addition of this interface is the lesser evil than having a small risk for potential issues due to the `any` cast below.

Similar to the `!` operator. In this case it is lying since there are cases where it can be undefined when using this class. I assume you didn't go for the `if` solution as a minor optimization, but even then there are type correct solutions possible such as having the buffered solution be the default option and the constructor taking a callback as input for example.

Reply:
> In this case it is lying since there are cases where it can be undefined when using this class.

But not actually: there's literally only one place where it is ever instantiated.

I'll see what I can do.

Reply:
~~Your suggested code won't work.~~ Actually we have late binding. Never mind!